### PR TITLE
Add fact to obtain iSCSI initiator IQN

### PR DESCRIPTION
**Pull request by sdodsley**
##### SUMMARY
This new fact enables the discovery of iSCSI initiator IQN value if a node has had iSCSI configured.
External storage modules can use this information to automate external storage array configuration
for hosts.
Fact discovery enabled for Linux, SunOS, AIX and Windows hosts.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
module_utils/facts/network

##### ANSIBLE VERSION
```
2.5
```

### CODE REVIEW

File: lib/ansible/module_utils/facts/system/iscsi.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+
+class IscsiInitiatorFactCollector(BaseFactCollector):
+    name = 'iscsi'
+    _fact_ids = set()
+
+    def collect(self, module=None, collected_facts=None):
+        iscsi_facts = {}
+        iscsi_facts['iscsi_iqn'] = {}
+        if sys.platform.startswith('linux') or sys.platform.startswith('sunos'):
```
Comment (alikins) [inline lib/ansible/module_utils/facts/system/iscsi.py]: Could you provide some sample versions of the text from these commands/files?

( /etc/iscsi/initiatorname.iscsi, `lsattr -E -l iscsi0` output, etc)

**Review by alikins — COMMENTED** [review]
Code looks pretty solid to me, but I wouldn't mind seeing some test data and test cases for IsciInitiatorFactCollector.collect()

File: lib/ansible/module_utils/facts/compat.py [hunk 2]
```diff
@@ -65,8 +65,8 @@ def ansible_facts(module, gather_subset=None):
     filter_spec = module.params.get('filter', '*')
 
     minimal_gather_subset = frozenset(['apparmor', 'caps', 'cmdline', 'date_time',
-                                       'distribution', 'dns', 'env', 'fips', 'local', 'lsb',
-                                       'pkg_mgr', 'platform', 'python', 'selinux',
+                                       'distribution', 'dns', 'env', 'fips', 'iscsi', 'local',
```
Comment (bcoca) [inline lib/ansible/module_utils/facts/compat.py]: `iscsi` does not belong in the minimal target, Hardware or Network are more appropriate

File: lib/ansible/module_utils/facts/system/iscsi.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+            for line in get_file_content('/etc/iscsi/initiatorname.iscsi', '').splitlines():
+                if line.startswith('#') or line.startswith(';') or line.strip() == '':
+                    continue
+                if line.startswith('InitiatorName='):
+                    iscsi_facts['iscsi_iqn'] = line.split('=', 1)[1]
+        elif sys.platform.startswith('aix'):
+            aixcmd = 'lsattr -E -l iscsi0 | grep initiator_name'
```
Comment (bcoca) [inline lib/ansible/module_utils/facts/system/iscsi.py]: use get_bin, this assumes lsattr is in path

**Review by bcoca — CHANGES_REQUESTED** [review]
not sure where it belongs but def not in minimal

**Comment by sdodsley** [comment]
@bcoca you mentioned using get_bin. I can't find this anywhere as a function in the repo. Can you confirm?

**Comment by bcoca** [comment]
sorry, its get_bin_path in module_utils/basic.py
https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py#L2213

**Comment by sdodsley** [comment]
@bcoca I went with just calling the exact path for the executable if that is OK. Seems to be the way it was done in the AIX hardware facts

**Comment by bcoca** [comment]
those need to be fixed also

**Comment by sdodsley** [comment]
@bcoca this OK for you now? Did I miss anything?

**Review by bcoca — APPROVED** [review]

**Comment by sdodsley** [comment]
@alikins Happy with this? I added examples of file contents and command string output as requested.

File: lib/ansible/module_utils/facts/network/iscsi.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+                if line.startswith('#') or line.startswith(';') or line.strip() == '':
+                    continue
+                if line.startswith('InitiatorName='):
+                    iscsi_facts['iscsi_iqn'] = line.split('=', 1)[1]
+        elif sys.platform.startswith('aix'):
+            aixcmd = '/usr/sbin/lsattr -E -l iscsi0 | grep initiator_name'
+            aixret = subprocess.check_output(aixcmd, shell=True)
+            if aixret[0].isalpha():
+                iscsi_facts['iscsi_iqn'] = aixret.split()[1].rstrip()
+        elif sys.platform.startswith('win'):
```
Comment (jborean93) [inline lib/ansible/module_utils/facts/network/iscsi.py]: This isn't needed as Windows hosts don't use the Python modules. If you wanted to add it into Windows have a look at https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/windows/setup.ps1

**Review by jborean93 — CHANGES_REQUESTED** [review]

**Review by jborean93 — APPROVED** [review]

**Comment by abadger** [comment]
Merged to devel for the 2.6 release series.

**Pull request merged**